Multi-column UDF support

richard-rogers opened the pull request:
## Description

Extends `DatasetSchema` with UDFs to compute new columns from existing columns (prior to resolving column metrics).

## Changes

* Adds new `UdfSchema` that supports dataset (multi-column) UDFs
* Register dataset UDFs with `@register_dataset_udf` decorator
* New privacy-preserving default submetric schema for `UdfMetric` 
* Deprecates `generate_udf_schema()` in favor of `generate_udf_resolvers()`

richard-rogers committed:
Experimental computed column support

FelipeAdachi reviewed (commented):
I personally like this approach better than the CallbackMetric

FelipeAdachi commented on python/tests/core/test_declarative_schema.py:
```diff
@@ -202,3 +203,29 @@ class UnknownType:
+def test_augmenters_row() -> None:
+    schema = DeclarativeSchema(
+        STANDARD_RESOLVER,
+        augmenters=[AugmenterSpec(column_name="col1", callbacks={"col2": lambda x: x, "col3": lambda x: x})],
```
what happens if the callback specifies the same col name (col1)? will it overwrite the original column?
what would be the ideal behavior in this case? don't allow it or let it overwrite? Or, maybe, create new cols, like `col1_aug_1` or something similar?

richard-rogers replied:
Good question :)

jamie256 replied:
Current implementation will overwrite the original input. We should at least emit an INFO log message that the derived column name exists in the input already.

jamie256 commented on python/whylogs/core/schema.py:
```diff
@@ -76,6 +85,7 @@ def __init__(
         schema_based_automerge: bool = False,
         segments: Optional[Dict[str, SegmentationPartition]] = None,
         validators: Optional[Dict[str, List[Validator]]] = None,
+        augmenters: List[AugmenterSpec] = [],
```
Move to an experimental subclass of DatasetSchema

jamie256 replied:
Default value here should not be an empty list (current implementation will create a shared instance across instances of DatasetSchema). Add default_factory here to get separate instances of the list for each DatasetSchema instance.

You may need to add imports for `field` and `default_factory`
```suggestion
        augmenters: List[AugmenterSpec] = field(default_factory=list),
```

jamie256 replied:
also renaming applies to the parameter here similarly to the class name (lets move away from augment*

jamie256 reviewed (commented):
Lets try these out a bit to see how they work with some of the scenarios we discussed around CV and NLP.

Minor comment about moving to experimental namespace while we are iterating on this API, but don't need to do that yet.

richard-rogers commented on python/whylogs/core/schema.py:
```diff
@@ -22,6 +22,15 @@
+class AugmenterSpec:
+    column_name: Optional[str] = None  # TODO: maybe make this a regex
+    column_type: Optional[Any] = None
+    callbacks: Dict[str, Callable[[Any], Any]] = field(
```
maybe `callbacks: List[Callable[[Any], Tuple[str, Any]]]` where the returned string is the new column name.

richard-rogers commented on python/whylogs/core/schema.py:
```diff
@@ -161,6 +173,30 @@ def _resolve_pdf(self, df: pd.DataFrame, force_resolve: bool = False) -> bool:
 
         return dirty
 
+    def _augment(
```
Maybe just make *this* the callback...

jamie256 replied:
This operation is a pre-profiling transform step, seems like it should live in Preprocessing conceptually. Having the pandas / row split in the signature is a leaky abstraction, it requires that the integrator know about this internal detail of how we split the inputs. Why not just pass the original data before the split, or we could pass them the extracted column where bound?

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

jamie256 commented on python/whylogs/core/schema.py:
```diff
@@ -22,6 +22,15 @@
 
 
+@dataclass
+class AugmenterSpec:
```
Suggest renaming the class: data augmentation is potentially a familiar term with respect to training ML models and we don't want to invoke that concept here.

What about naming it something like: Transforms, Transformer, ColumnTransformer, DerivedColumnSpec, DerivedColumn?

jamie256 commented on python/whylogs/core/schema.py:
```diff
@@ -161,6 +173,30 @@ def _resolve_pdf(self, df: pd.DataFrame, force_resolve: bool = False) -> bool:
+                for augmenter in self.augmenters:
+                    col_name, col_type = augmenter.column_name, augmenter.column_type
+                    why_type = self.type_mapper(type(value))
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```
Factor this out into separate function. Maybe this is a transforms resolve step?

You can simplify` (col_name and col_name == column)` to `(col_name == column)`

jamie256 commented on python/whylogs/core/schema.py:
```diff
@@ -161,6 +173,30 @@ def _resolve_pdf(self, df: pd.DataFrame, force_resolve: bool = False) -> bool:
+                for augmenter in self.augmenters:
+                    col_name, col_type = augmenter.column_name, augmenter.column_type
+                    why_type = pandas.dtypes[column]
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```
same comment as above, also will let us avoid the duplicated if statement.

jamie256 commented on python/whylogs/core/schema.py:
```diff
@@ -161,6 +173,30 @@ def _resolve_pdf(self, df: pd.DataFrame, force_resolve: bool = False) -> bool:
+                    why_type = pandas.dtypes[column]
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
+                        for new_col, callback in augmenter.callbacks.items():
+                            pandas[new_col] = pandas[column].map(callback)
```
Let not add the derived columns to the input pandas dataframe which modifies the input. Can we keep these in a separate dataframe?

jamie256 reviewed (commented):
Needs to be moved into experimental and not break/add to existing core classes yet to protect integrations from accidentally relying on this experimental extensibility point.

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
review feedback, experimental namespace

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
pre-commit

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,83 @@
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Mapping, Optional, Tuple
+
+from whylogs.core.datatypes import TypeMapper
```
```suggestion
from whylogs.core.datatypes import DataType, TypeMapper
```

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,83 @@
+@dataclass
+class UdfSpec:
+    column_name: Optional[str] = None  # TODO: maybe make this a regex
+    column_type: Optional[Any] = None
```
```suggestion
    column_type: Optional[DataType] = None
```

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,83 @@
+                for udf_spec in self.udf_specs:
+                    col_name, col_type = udf_spec.column_name, udf_spec.column_type
+                    why_type = self.type_mapper(type(value))
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```
```suggestion
                    if col_name == column or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,83 @@
+                for udf_spec in self.udf_specs:
+                    col_name, col_type = udf_spec.column_name, udf_spec.column_type
+                    why_type = pandas.dtypes[column]
+                    if (col_name and col_name == column) or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```
```suggestion
                    if col_name == column or (col_name is None and isinstance(col_type, why_type)):  # type: ignore
```

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers committed:
review feedback

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,89 @@
+            segments=segments,
+            validators=validators,
+        )
+        self.udf_specs = udf_specs or []
```
```suggestion
        self.udf_specs = deepcopy(udf_specs) or []
```

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,89 @@
+
+    def _run_udfs_on_row(self, row: Mapping[str, Any], new_columns: Mapping[str, Any]) -> None:
+        for column, value in row.items():
+            for udf_spec in self.udf_specs:
```
Change this to map from column name -> UDF list and map from column type -> UDF list for efficiency

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
work

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
fix tests

richard-rogers committed:
debugging

richard-rogers committed:
test additional udf specs

richard-rogers committed:
index type UDFs by string instead of class

richard-rogers committed:
schema ctor arg pass thru

richard-rogers committed:
multi-column UDFs working

richard-rogers committed:
metric pass-thru

richard-rogers committed:
docstr updates and try/except

richard-rogers committed:
review feedback

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
dataset UDFs are all multicolumn; no type matching; better-ish exception handling

richard-rogers committed:
allow empty resolver list

richard-rogers committed:
allow empty resolvers

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

richard-rogers committed:
support UdfMetric registration in dataset UDF schema

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers committed:
pre-commit

jamie256 commented on python/whylogs/experimental/core/metrics/udf_metric.py:
```diff
@@ -305,6 +305,11 @@ def upper(x):
     return resolvers
 
 
+@deprecated(message="Please use generate_udf_resovlers()")
```
```suggestion
@deprecated(message="Please use generate_udf_resolvers()")
```

jamie256 reviewed (dismissed):
Left some minor comments.

richard-rogers committed:
Update python/whylogs/experimental/core/metrics/udf_metric.py

Co-authored-by: Jamie Broomall <88007022+jamie256@users.noreply.github.com>

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,208 @@
+    def _run_udfs_on_row(self, row: Mapping[str, Any], new_columns: Dict[str, Any]) -> None:
+        for spec in self.multicolumn_udfs:
+            if set(spec.column_names).issubset(set(row.keys())):
+                _apply_udfs_on_dataframe(row, spec.udfs, new_columns)
```
```suggestion
                _apply_udfs_on_row(row, spec.udfs, new_columns)
```

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,208 @@
+) -> UdfSchema:
+    """
+    Returns a UdfSchema that implements any registered UDFs, along with any
+    other_udf_specs or resovlers passed in.
```
```suggestion
    other_udf_specs or resolvers passed in.
```

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

richard-rogers committed:
Update python/whylogs/experimental/core/udf_schema.py

jamie256 committed:
Have UDF base schema not include frequent items by default (#1244)


We need an easier calling pattern for integrations using the UDF schemas
that are privacy preserving. Here's one way we could support that.

Co-authored-by: richard-rogers <93153899+richard-rogers@users.noreply.github.com>

richard-rogers committed:
Merge branch 'mainline' into dev/richard/augment

jamie256 reviewed (approved):
LGTM!

richard-rogers commented on python/whylogs/experimental/core/udf_schema.py:
```diff
@@ -0,0 +1,210 @@
+        try:
+            new_df[new_col] = udf(pandas)
+        except Exception:  # noqa
+            # new_df[new_col] = pandas[pandas.keys()[0]].apply(lambda x: None)
```
column with single `None`

jamie256 reviewed (dismissed):
LGTM!

richard-rogers committed:
bug fix; update exception handling tests
